### PR TITLE
[DebugInfo][RemoveDIs] Emulate inserting insts in dbg.value sequences

### PR DESCRIPTION
**Pull request by jmorse**
Here's a problem for the RemoveDIs project to make debug-info not be stored in instructions -- in the following sequence:
    dbg.value(foo
    %bar = add i32 ...
    dbg.value(baz
It's possible for rare passes (only CodeGenPrepare) to remove the add instruction, and then re-insert it back in the same place. When debug-info is stored in instructions and there's a total order on "when" things happen this is easy, but by moving that information out of the instruction stream we start having to do manual maintenance.

This patch adds some utilities for re-inserting an instruction into a sequence of DPValue objects. Someday we hope to design this away, but for now it's necessary to support all the things you can do with dbg.values. The two unit tests show how DPValues get shuffled around using the relevant function calls. A follow-up patch adds instrumentation to CodeGenPrepare.

### CODE REVIEW

**Comment by jmorse** [comment]
More mucking around with the CodeGenPrepare patch (uploading shortly) has revealed that actually the supported use case is where the reinserted instruction is placed with `insertAfter` rather than insertBefore, which changes the set of inputs that needs to be supported. The commit just pushed switches this facility to support that (and udpate the tests for it).

I feel it's alright to support this limited situation seeing how it's only for one consumer (CodeGenPrepare). One day codegenprepare will go away anyway... according to the file comment... written in 2008.

File: llvm/lib/IR/DebugProgramInstruction.cpp [hunk 1]
```diff
@@ -267,6 +267,19 @@ void DPMarker::dropOneDPValue(DPValue *DPV) {
   DPV->deleteInstr();
 }
 
+std::optional<DPValue::self_iterator> DPMarker::getReinsertionPosition() {
```
Comment (OCHyams) [inline llvm/lib/IR/DebugProgramInstruction.cpp:270]: Does this need to be a `DPMarker` method? It doesn't look like any members are used other than `MarkedInstr`, which I imagine can be found easily at call sites.

The use-case in #73660 looks like it creates a DPMarker just to call this function (I could be misreading it though?).

Comment (jmorse) [reply llvm/lib/IR/DebugProgramInstruction.cpp:270]: Uuuuhhhh. Yeah, I think it does.

It's really not clear where this method should be placed -- I think it was in originally in BasicBlock, then I moved it here.... how would you feel about it being back in BasicBlock?

Comment (OCHyams) [reply llvm/lib/IR/DebugProgramInstruction.cpp:270]: Yeah that sounds ok to me. Or can it be standalone (if so, maybe a static method on DPMarker makes sense?)? Unsure.

**Review by OCHyams — APPROVED** [review]
LGTM

**Pull request merged**